File: apps/memory-lab-py/memory_lab_py/l2_builder.py
from __future__ import annotations

from .utils_id import build_l2_project_index_id, build_l2_time_index_id, now_iso


def _parse_date_key(time_period: str) -> str:
    idx = time_period.find(":")
    return time_period[:idx] if idx >= 0 else time_period


def _derive_project_status(summary: str, skills: dict) -> str:
    lower = summary.lower()
    for rule in skills["projectStatusRules"]["rules"]:
        for keyword in rule["keywords"]:
            if keyword.lower() in lower:
                return rule["status"]
    return skills["projectStatusRules"]["defaultStatus"]


def build_l2_time_from_l1(l1: dict) -> dict:
    date_key = _parse_date_key(l1["timePeriod"])
    now = now_iso()
    return {
        "l2IndexId": build_l2_time_index_id(date_key),
        "dateKey": date_key,
        "summary": l1["summary"],
        "l1Source": [l1["l1IndexId"]],
        "createdAt": now,
        "updatedAt": now,
    }


def build_l2_projects_from_l1(l1: dict, skills: dict) -> list[dict]:
    now = now_iso()
    result: list[dict] = []
    for project_name in l1["projectTags"]:
        result.append(
            {
                "l2IndexId": build_l2_project_index_id(project_name),
                "projectName": project_name,
                "summary": f"{project_name}：{l1['summary']}",
                "currentStatus": _derive_project_status(l1["summary"], skills),
                "latestProgress": l1["situationTimeInfo"],
                "l1Source": [l1["l1IndexId"]],
                "createdAt": now,
                "updatedAt": now,
            }
        )
    return result

File: packages/openclaw-memory-plugin/src/core/indexers/l2-builder.ts
import type { L1WindowRecord, L2ProjectIndexRecord, L2TimeIndexRecord } from "../types.js";
import { buildL2ProjectIndexId, buildL2TimeIndexId, nowIso } from "../utils/id.js";

function parseDateKey(timePeriod: string): string {
  const idx = timePeriod.indexOf(":");
  return idx >= 0 ? timePeriod.slice(0, idx) : timePeriod;
}

export function buildL2TimeFromL1(l1: L1WindowRecord): L2TimeIndexRecord {
  const dateKey = parseDateKey(l1.timePeriod);
  const now = nowIso();
  return {
    l2IndexId: buildL2TimeIndexId(dateKey),
    dateKey,
    summary: l1.summary,
    l1Source: [l1.l1IndexId],
    createdAt: now,
    updatedAt: now,
  };
}

export function buildL2ProjectsFromL1(l1: L1WindowRecord): L2ProjectIndexRecord[] {
  const now = nowIso();
  return l1.projectDetails.map((project) => ({
    l2IndexId: buildL2ProjectIndexId(project.key),
    projectKey: project.key,
    projectName: project.name,
    summary: project.summary || `${project.name}: ${l1.summary}`,
    currentStatus: project.status,
    latestProgress: project.latestProgress || l1.situationTimeInfo,
    l1Source: [l1.l1IndexId],
    createdAt: now,
    updatedAt: now,
  }));
}
